docs/apache-airflow/security/secrets/secrets-backend/index.rst

```diff
@@ -101,8 +143,8 @@ Roll your own secrets backend
 A secrets backend is a subclass of :py:class:`airflow.secrets.BaseSecretsBackend` and must implement either
 :py:meth:`~airflow.secrets.BaseSecretsBackend.get_connection` or :py:meth:`~airflow.secrets.BaseSecretsBackend.get_conn_value`.
 
-After writing your backend class, provide the fully qualified class name in the ``backend`` key in the ``[secrets]``
-section of ``airflow.cfg``.
+After writing your backend class, provide the fully qualified class name in the ``backend`` key or provide
```

Comment:
I agree that better support only one way of configuration secrets backends.
However I think old way in any cases should be supported until it entirely remove from Airflow (3.0.0 ?)

Reply:
right and what i am telling you is that you need to initiate deprecation properly, using the framework we have for this kind of thing.   for example see here: https://github.com/apache/airflow/blob/main/airflow/configuration.py#L171

the option for `base_log_folder` was moved from `core` to `logging` -- we don't keep the deprecated values around in the old location.  we only keep the new location.  and we handle "upgrading" deprecated approaches in AirflowConfigParser.  see also methods `_upgrade_auth_backends` and `_upgrade_postgres_metastore_conn`.   so it seems to me the right approach is to formally remove it from the config immediately as a deprecation and add the logic to "upgrade" or "convert" the deprecated legacy values.

Reply:
Approach in general fine only one thing that could be a problem to do it in the same way for secrets backend

https://github.com/apache/airflow/blob/b692517ce3aafb276e9d23570e9734c30a5f3d1f/airflow/configuration.py#L1600-L1602

Secrets Backend initialised before config validated

Reply:
This probably it not a big deal actually. Need to check it a bit later. 
Seems like when secret backends initialised first time it can't get any values, so it initialised with default backend

Reply:
There are quite a few findings with use `deprecated_values` and `upgraded_values` for Secrets Backends

1. We need to deprecate in once two configs `[secrets] backend` and `[secrets] backend_kwargs` by `[secrets] backends_config`. I think it is not possible right now.

2. Current and new config expected different types. New JSON List, Current just a string. So warning would be 
```airflow/airflow/configuration.py:552: DeprecationWarning: The backend option in [secrets] has been renamed to backends_config - the old setting has been used, but please update your config.```

3. We want to upgrade config so it should be done before
https://github.com/apache/airflow/blob/b692517ce3aafb276e9d23570e9734c30a5f3d1f/airflow/configuration.py#L1601-L1602

4. One side effect about transformation old value to new one, we need to call config.get(section='secrets', key='backends_config') and we don't know where this value stored in `backend` or in `backends_config`. 

I would rather for now add Depreciation warning in `initialize_secrets_backends()` and after think about how to solve all findings which listed above
https://github.com/apache/airflow/blob/68e0741b8cd0554948be4db7946d8c7afaa5b4d0/airflow/configuration.py#L1596-L1600

Reply:
> I would rather for now add Depreciation warning in initialize_secrets_backends() and after think about how to solve all findings which listed above

i understand the impulse but this does not make sense to me. it is hard to think through backcompat and deprecation and UX -- this is one of the challenges with open source.  we should not have two ways of doing things and should not defer til "later" these problems.  the right time to deal with deprecation of an "old" way of configuration is when you add the "new" way of configuration.

Reply:
Secrets Backends itself this is exactly "two ways of doing things" - it initialised outside of configuration and it ignores all validations, most possible for avoid Chicken or the egg. So it is already kind of exception in airflow.configurations

The idea of this PR extends current implementation Secrets Backend - make it more flexible, not rewrite entire configurations of Airflow and deprecation model.

My current proposal it is focus on implementation of Advanced Secrets Backend rather that try solve all issues of configurations related only to SB. Or you think that this actually blocker?

Reply:
I'm not proposing `rewrite entire configurations of Airflow and deprecation model` or `solve all issues of configurations related only to SB`.  I'm essentially saying, when you want to change configuration style, you _change_ it; you don't add a _another_ style.

I understand that with SB there are complexities with being also able to take "secret" config items from it, and I understand that unfortunately it's a bit chicken egg and confusing and it's not my favorite --- but it's already like that.  So maybe this means you can't copy the existing "upgrade" examples exactly -- maybe you have to do something a little different, a little more difficult.  But the fundamental principle here is not to "multiply" configuration approaches but to just _change_ the configuration approach, and to maintain backcompat with deprecation warning for a time.  I am pretty confident you can do that.

But keep in mind, I'm not the boss, right?  This is open source.  And if you disagree with my thinking, and if no one else is chimes in here, one thing you can always do is start a discussion on the dev list, starting a thread like `[DISCUSS] To add or update secrets backend configuration approach` and you can state your case and see if you find support.

Reply:
Nothing personal but let me explain how it looks like for my side. I invest some my time to check how it possible to do it right now with same approach and found that withouts hacks in runtime it is hardly possible but you've told me:

> So maybe this means you can't copy the existing "upgrade" examples exactly -- maybe you have to do something a little different, a little more difficult

Which more like for me `I don't know how to do it, but think better and do it better` - that is last thing that I or someone else expects see on review (both on open source or commercial development) . If you have suggestion how to do it better right now without rewrite a lot of stuff - feel free to suggest.

---

Right now Secrets Backends ignore "better approach", implicit upgrades values and do not track it. I think everyone sleep well

https://github.com/apache/airflow/blob/2936759256273a66cf6343e451d81013a95ec47e/airflow/configuration.py#L1506-L1510

---

Just for reduce your or someone else time for investigate how to do it "in right way right now".

### Specify directly AirflowConfigParser.deprecated_options

You can't just placed in AirflowConfigParser.deprecated_options something like `('secrets', 'backends_config'): ('secrets', 'backend', '2.4.0')` because we need to check that **[secrets] backend** defined and upgrades required.

```python

    def _upgrade_secrets_backend(self):
        old_backend = self.get(section='secrets', key='backend', fallback='')
        if old_backend:
            ...
```

Code above always will return warning even if user do not define anything in **[secrets] backend**
_DeprecationWarning: The backend option in [secrets] has been renamed to backends_config - the old setting has been used, but please update your config._

_Possible solution?_ Define in deprecation config `('secrets', 'backends_config'): ('secrets', 'backend', '2.4.0')` right after upgrade. Which might be unclear for developers, because they should keep it in their mind that something outside could change AirflowConfigParser.deprecated_options

### Deprecate [secrets] backend_kwargs

It is not possible right now by define something in AirflowConfigParser.

_Possible solution?_ Show warning that this config doesn't use anymore and anywhere on upgrade. However right now we also could warn users with custom message on init secrets backend

### Waring show that only one option deprecated

We need to inform user by this message
```
DeprecationWarning: The backend and backend_kwargs options in [secrets] has been moved to backends_config - the old setting has been used, but please update your config.
```

But now we can get only predefined
```
            warnings.warn(
                f'The {deprecated_name} option in [{deprecated_section}] has been moved to the {key} option '
                f'in [{section}] - the old setting has been used, but please update your config.',
                DeprecationWarning,
                stacklevel=3,
            )
```

_Possible solution?_ Rewrite AirflowConfigParser.deprecated_options and methods which check and show this warning to make it more generic.

---

I'm absolutely agree with your idea that we need to have only one approach with deprecated values.
But right now Secrets Backend doesn't use this approach in anyway and without changes in AirflowConfigParser it is hardly possible.

And this PR only about extends functional of Secrets Backend not changes in AirflowConfigParser. If we want to change AirflowConfigParser better to open separate PR in this case if we need to revert back changes we would revert "Add advanced secrets backend configurations" or "Make AirflowConfigParser more generic with deprecation  warning" but not both.

Reply:
> Which more like for me I don't know how to do it, but think better and do it better - that is last thing that I or someone else expects see on review (both on open source or commercial development) . If you have suggestion how to do it better right now without rewrite a lot of stuff - feel free to suggest.

I'm just saying on the face of it, it seems it should be possible to update the config approach for secrets backend without keeping the old one live.  And no, I did not investigate exactly how you should do this.

> I'm absolutely agree with your idea that we need to have only one approach with deprecated values.
But right now Secrets Backend doesn't use this approach in anyway and without changes in AirflowConfigParser it is hardly possible.

What i'm saying is that we should not maintain two config styles concurrently, that the old one should be deprecated, and the nuts and bolts of exactly how I'm less concerned with.

I may be able to explore solutions next week.  In the meantime who knows maybe someone else will support your approach. 

Thanks

Reply:
Thanks for suggestion!
I don't think I got how it actually work but I need check also this way.